Data structure solution

olegovich22 opened the pull request:
Was created a simple interface for the linked list using sysfs attributes. 
The interface has 4 cmd: add tail, add head, rm tail, and rm head.

ekovalyov committed:
homework: basic data structures

olegovich22 committed:
Data structure: add sysfs interface for list

Was created simple interface for linked
list using sysfs attributes. Interface
has 4 cmd: add tail, add head, rm tail
and re head.

Signed-off-by: Oleh Yakymenko <o.yakymenko22@gmail.com>

ekovalyov commented on 04_basic_struct/hello.c:
```diff
@@ -0,0 +1,240 @@
+					GFP_KERNEL);
+
+	if (node == NULL)
+		pr_err("Element hasn`t created. Not enaugh memory for list entry\n");
```
Typing error in word enough

ekovalyov commented on 04_basic_struct/hello.c:
```diff
@@ -0,0 +1,240 @@
+	if (a_node == NULL)
+		return;
+
+	if (a_node->data != NULL)
+		kfree(a_node->data);
```
If obj is NULL, no operation is performed. So the condition check is not required.

ekovalyov commented on 04_basic_struct/hello.c:
```diff
@@ -0,0 +1,240 @@
+		if (buf_offset + arrow_size + 1 > PAGE_SIZE)
+			break;
+
+		sprintf(buf+buf_offset, "%s", current_node->data);
+		buf_offset += strlen(current_node->data)-1;
+		sprintf(buf+buf_offset, "%s", arrow);
+		buf_offset += arrow_size;
```
You can use one sprintf and use its return value for buffer_offset modification.

ekovalyov commented on 04_basic_struct/hello.c:
```diff
@@ -0,0 +1,240 @@
+			break;
+
+		sprintf(buf+buf_offset, "%s", current_node->data);
+		buf_offset += strlen(current_node->data)-1;
```
For sprintf. If successful, the total number of characters written is returned excluding the null-character appended at the end of the string, otherwise a negative number is returned in case of failure.

ekovalyov commented on 04_basic_struct/Makefile:
```diff
@@ -0,0 +1,11 @@
+ifneq ($(KERNELRELEASE),)
+# kbuild part of makefile
+obj-m := hello.o
```
Please add indentation for this line.

ekovalyov commented on 04_basic_struct/hello.c:
```diff
@@ -0,0 +1,240 @@
+			"ah_<string> - add to head\n"
+			"rt - remove from tail\n"
+			"rh - remove from head");
+		return strlen(buf);
```
You can just return result from sprintf

ekovalyov reviewed (changes requested):
Good job. Please fix mentioned minor notes.

olegovich22 committed:
Data structure: improve efficiency for print fun

Was improved efficiency for list print function.
Fixed useless strlen counting. Fixed some minor
issues.

Signed-off-by: Oleh Yakymenko <o.yakymenko22@gmail.com>

olegovich22 commented:
Fixed
